src/diffusion_for_multi_scale_molecular_dynamics/noisers/lattice_noiser.py

```diff
@@ -1,23 +1,101 @@
+            shape : the shape of the sample.
 
         Returns:
-            real_lattice_vectors: a sample of noised lattice vectors. Placeholder for now.
+            gaussian_noise: a sample from N(0, 1) of dimensions shape.
         """
-        return real_lattice_vectors
+        return torch.randn(shape)
+
+    def get_noisy_lattice_vectors(
+        self,
+        real_lattice_parameters: torch.Tensor,
+        sigmas_n: torch.Tensor,
+        alpha_bars: torch.Tensor,
+        num_atoms: torch.Tensor,
+    ) -> torch.Tensor:
+        r"""Get noisy lattice vectors.
+
+        We consider the lattice parameters as a tensor of spatial_dimension degrees of freedom i.e. we assume the unit
+        cell is orthogonal.
+
+        .. math::
+
+            q(L_t | L_0) = \mathcal{N}\left(\sqrt{\bar{\alpha}_t}L_0 +
+                (1 - \sqrt{\bar{\alpha}_t})\mu(n)I, (1-\sqrt{\bar{\alpha}_t})\sigma^2_t(n)I\right)
+
+        Args:
+            real_lattice_parameters: lattice parameters from the sampled data. These parameters are not the lattice
+                vector, but an array of dimension [spatial_dimension] containing the size of the orthogonal box.
+                TODO add angles
+            sigmas_n: variance of the perturbation kernel rescaled by the number of atoms. Tensor is assumed to be of
+                the same shape as real_lattice_parameters.
+            alpha_bars: cumulative noise scale. Tensor is assumed to be of the same shape as real_lattice_parameters.
+            num_atoms: number of atoms in each sample. Tensor should be a 1D tensor matching the first dimension of the
+                real_lattice_parameters tensor.
+
+        Returns:
+            noisy_lattice_parameters: a sample of noised lattice parameters as tensor of size [spatial_dimension].
+        """
+        # TODO add angles
+        assert (
+            real_lattice_parameters.shape == sigmas_n.shape
+        ), "sigmas array is expected to be of the same shape as the real_lattice_parameters array"
+
+        assert (
+            alpha_bars.shape == sigmas_n.shape
+        ), "sigmas array is expected to be of the same shape as the alpha_bars array"
+
+        z_scores = self._get_gaussian_noise(real_lattice_parameters.shape).to(
+            sigmas_n
+        )
+
+        noise_width = (1 - alpha_bars) * sigmas_n**2
+        sqrt_alpha_bars = torch.sqrt(alpha_bars)
+        average_density = torch.zeros_like(num_atoms)
+        average_density[:, :self.spatial_dimension] = torch.pow(
+            num_atoms[:, :self.spatial_dimension],
+            1 / self.spatial_dimension,
+        )  / self.inverse_density
+
+        noisy_lattice_parameters_avg = (
+            sqrt_alpha_bars * real_lattice_parameters
+            + (1 - sqrt_alpha_bars) * average_density
+        )
+        # we are not limiting the range of value to anything
+        noisy_lattice_parameters = noise_width * z_scores + noisy_lattice_parameters_avg
```

Comment:
I think you are implementing this equation from MatterGen:
<img width="845" alt="Screenshot 2024-12-24 at 9 45 56 AM" src="https://github.com/user-attachments/assets/d847632c-27c5-46d9-aeaf-575665f33945" />

A sample from a Gaussian $N(\mu, \sigma^2)$ should go like $x = \mu + \sigma z$ where $z\sim N(0, 1)$. However, you are multiplying by `noisy_width`, which is like `\sigma^2`.

Reply:
good catch - fixed